src/llscan.cc

```diff
@@ -0,0 +1,376 @@
+          "No memory range information available for this process. Cannot scan "
+          "for objects.\n");
+      return false;
+    }
+  }
+
+  /* If we've reached here we have access to information about the valid memory
+   * ranges in the
+   * process and can scan for objects.
+   */
+
+  /* Populate the map of objects. */
+  if (mapstoinstances_.empty()) {
+    FindJSObjectsVisitor v(target, GetMapsToInstances());
+
+    ScanMemoryRanges(v);
+  }
+
+  return true;
+}
+
+
+void LLScan::ScanMemoryRanges(FindJSObjectsVisitor& v) {
+  MemoryRange* head = ranges_;
+
+  bool done = false;
+
+  while (head != nullptr && !done) {
+    uint64_t address = head->start;
+    uint64_t len = head->length;
+    head = head->next;
+
+    /* Brute force search - query every address - but allow the visitor code to
+     * say
+     * how far to move on so we don't read every byte.
+     */
+    for (auto searchAddress = address; searchAddress < address + len;) {
+      uint32_t increment =
+          v.Visit(searchAddress, (address + len) - searchAddress);
+      if (increment == 0) {
+        done = true;
+        break;
+      }
+      searchAddress += increment;
+    }
+  }
+}
+
+
+/* Read a file of memory ranges parsed from the core dump.
+ * This is a work around for the lack of an API to get the memory ranges
+ * within lldb.
+ * There are scripts for generating this file on mac and linux stored as a
+ * snippet on github:
+ * https://gist.github.com/hhellyer/6d79009e142d6eef696525afe1ecea43
+ * export the name or full path to the ranges file in the RANGESFILE env var
+ * before starting
+ * lldb and loading the llnode plugin.
+ */
+bool LLScan::GenerateMemoryRanges(lldb::SBTarget target,
+                                  const char* segmentsfilename) {
+  std::ifstream input(segmentsfilename);
+
+  if (!input.is_open()) {
+    return false;
+  }
+
+  uint64_t address = 0, len = 0;
+
+  MemoryRange** tailptr = &ranges_;
+
+  lldb::addr_t address_byte_size = target.GetProcess().GetAddressByteSize();
+
+  while (input >> std::hex >> address >> std::hex >> len) {
+    /* Check if the range is accessible.
+     * The structure of a core file means if you check the start and the end of
+     * a range then the
+     * middle will be there, ranges are contiguous in the file, but cores often
+     * get truncated due
+     * to file size limits so ranges can be missing or truncated. Sometimes
+     * shared memory segments
+     * are omitted so it's also possible an entire section could be missing from
+     * the middle.
+     */
+    lldb::SBError error;
+    target.GetProcess().ReadPointerFromMemory(address, error);
+    if (error.Success()) {
+      target.GetProcess().ReadPointerFromMemory(
+          (address + len) - address_byte_size, error);
+
+      if (error.Success()) {
+        MemoryRange* newRange = new MemoryRange();
+        if (nullptr == newRange) {
+          return false;
+        }
+        newRange->start = address;
```

Comment:
Done, I haven't gone all the way and added accessors for all the fields of MemoryRange as the class is already a private inner class of LLScan - effectively part of the LLScan implementation.